File: src/dataset/javascript/saddle-points_exercism-javascript-ThomasZumsteg/gemini/GeminiFlash_saddle-points_v3.js
var Matrix = function(matrix) {
	/* Two dimentional matrix class */
	const rows = matrix.split("\n").map(row => {
		return row.trim().split(/\s+/).map(e => parseInt(e, 10));
	});

	const numCols = rows[0].length;
	const columns = Array.from({ length: numCols }, (_, col) =>
		rows.map(row => row[col])
	);

	this.rows = rows;
	this.columns = columns;
	this.saddlePoints = saddlePoints(this);
}

function saddlePoints(matrix) {
	/* Finds saddle points in a matrix */
	const rows = matrix.rows;
	const columns = matrix.columns;
	const numRows = rows.length;
	const numCols = columns.length;

	const rowMax = rows.map(row => Math.max(...row));
	const colMin = columns.map(col => Math.min(...col));

	const saddlePoints = [];

	for (let r = 0; r < numRows; r++) {
		const row = rows[r];
		for (let c = 0; c < numCols; c++) {
			const element = row[c];
			if (element === rowMax[r] && element === colMin[c]) {
				saddlePoints.push([r, c]);
			}
		}
	}

	return saddlePoints;
};

export default Matrix;

File: src/dataset/go/saddle_points_Exercism_go_ThomasZumsteg/gemini/GeminiFlash_saddle_points_v3.go
package matrix

// Pair is a row and column in a matrix
type Pair struct {
	row, col int
}

/*Saddle finds saddle points in a matrix.*/
func (m *Matrix) Saddle() []Pair {
	var saddlePoints []Pair
	rows := m.Rows()
	cols := m.Cols()
	numRows := len(rows)
	numCols := len(cols)

	for r := 0; r < numRows; r++ {
		row := rows[r]
		for c := 0; c < numCols; c++ {
			elem := row[c]

			isSaddle := true

			// Check if element is smallest in row
			for _, val := range row {
				if val < elem {
					isSaddle = false
					break
				}
			}

			if !isSaddle {
				continue
			}

			// Check if element is largest in column
			for i := 0; i < numRows; i++ {
				if rows[i][c] > elem {
					isSaddle = false
					break
				}
			}

			if isSaddle {
				saddlePoints = append(saddlePoints, Pair{r, c})
			}
		}
	}
	return saddlePoints
}